### PR TITLE
Feature/use document visibility

### PR DESCRIPTION
**Pull request by Artikool5**
[Codesendbox](https://codesandbox.io/p/devbox/quiet-thunder-v7xg6x?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clyh2lwaz00062v6jltbpzxey%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clyh2lway00022v6jnwuijzj3%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clyh2lway00042v6js1t937hr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clyh2lway00052v6jmit984xf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clyh2lway00022v6jnwuijzj3%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clyh2lwax00012v6jz7uqcquv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clyh2lway00022v6jnwuijzj3%2522%252C%2522activeTabId%2522%253A%2522clyh2lwax00012v6jz7uqcquv%2522%257D%252C%2522clyh2lway00052v6jmit984xf%2522%253A%257B%2522id%2522%253A%2522clyh2lway00052v6jmit984xf%2522%252C%2522activeTabId%2522%253A%2522clyh3vi0400262v6ju0fhj6l9%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A5173%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clyh3vi0400262v6ju0fhj6l9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clyh2lway00042v6js1t937hr%2522%253A%257B%2522id%2522%253A%2522clyh2lway00042v6js1t937hr%2522%252C%2522activeTabId%2522%253A%2522clyh2y9d1000q2v6jikqxij3p%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clyh2lway00032v6jxxbop6kp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clyh2ly7l000kdjh03hs42hne%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clyh2y9d1000q2v6jikqxij3p%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

[npm пакет](https://www.npmjs.com/package/@artikool5/use-document-visibility)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,15 +1,35 @@
...
+  "devDependencies": {
+    "@rollup/plugin-typescript": "^11.1.6",
+    "@types/react": "^18.3.3",
+    "rimraf": "^6.0.0",
+    "rollup": "^4.18.1",
+    "tslib": "^2.6.3",
+    "typescript": "^5.5.3"
+  },
+  "dependencies": {
+    "react": "^18.3.1"
```
Comment (heyanniemoran) [inline package.json]: Нужно давать возможность пользоваться пакетом, если версия реакта поддерживает хуки. Т.е. поменять бы на более раннюю

File: src/main.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+type VisibilityListener = (isVisible: boolean) => void;
+
+interface DocumentVisibilityReturn {
+  isVisible: boolean;
+  count: number;
+  onVisibilityChange: (listener: VisibilityListener) => () => void;
+}
+
+export const useDocumentVisibility = (): DocumentVisibilityReturn => {
+  const [isVisible, setIsVisible] = useState(!document.hidden);
```
Comment (heyanniemoran) [inline src/main.ts]: в ssr не будет работать

**Comment by heyanniemoran** [comment]
readme бы добавить
image

File: src/main.ts (hunk 2 above)
Comment (Artikool5) [reply src/main.ts]: Извини, не разбираюсь в ssr. В какую сторону копать?

Comment (Artikool5) [reply src/main.ts]: Блин, я все на экпорт смотрел, а ты наверное про обращение к document на сервере 😅

**Review by heyanniemoran — APPROVED** [review]

**Review by Mary-fox — APPROVED** [review]

**Comment by heyanniemoran** [comment]
Можно мержить

**Pull request merged**